google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/PostgreSQLStatementParser.java

```diff
@@ -149,30 +152,76 @@ ParametersInfo convertPositionalParametersToNamedParametersInternal(char paramCh
     return new ParametersInfo(paramIndex - 1, named.toString());
   }
 
-  private int skip(String sql, int currentIndex, StringBuilder result) {
+  /**
+   * Note: This is an internal API and breaking changes can be made without prior notice.
+   *
+   * <p>Returns the PostgreSQL-style query parameters ($1, $2, ...) in the given SQL string. The
+   * SQL-string is assumed to not contain any comments. Use {@link #removeCommentsAndTrim(String)}
+   * to remove all comments before calling this method. Occurrences of query-parameter like strings
+   * inside quoted identifiers or string literals are ignored.
+   *
+   * <p>The following example will return a set containing ("$1", "$2"). <code>
+   * select col1, col2, "col$4"
+   * from some_table
+   * where col1=$1 and col2=$2
+   * and not col3=$1 and col4='$3'
+   * </code>
+   *
+   * @param sql the SQL-string to check for parameters. Must not contain comments.
+   * @return A set containing all the parameters in the SQL-string.
+   */
+  @InternalApi
+  public Set<String> getQueryParameters(String sql) {
```

Comment:
I imagine we don't need to consider malformed queries here (`select col$4" from some_table ...`)

Reply:
Correct. That is caught by the preceding step (`removeCommentsAndTrim`) that should be taken by any well-informed users of the API.